### PR TITLE
fix: torch backend- gather fix

### PR DESCRIPTION
**Pull request by Kacper-W-Kozdon**
# PR Description

Previous implementation of gather() from torch backend was incorrect but did not get caught by pytest. I corrected a vast majority of issues. **_Pytest might still show some errors for edge cases due to the very limited implementation of torch.gather() and various workarounds needed to recreate tensorflow's ground truth behaviour, they'll likely be related to dimensionality and I'm still removing them (it's all in the conditions for reshaping and expanding)_**. Side note: the official tensorflow's documentation for gather contains errors in the expression to calculate the dimensions of the output.

## Related Issue

Closes https://github.com/unifyai/ivy/issues/26498

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you run your tests and are your tests passing?
- [x] Did pre-commit not fail on any check?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by Kacper-W-Kozdon** [comment]
## Reformatting Task Checklist
#### IMPORTANT NOTICE 🚨:
The [Ivy Docs](https://unify.ai/docs/ivy/) represent the ground truth for the task descriptions and this checklist should only be used as a supplementary item to aid with the review process.

#### LEGEND 🗺:
- ❌ :  Check item is not completed.
- ✅ :  Check item is ready for review.
- 🆘 :  Stuck/Doubting implementation (PR author should add comments explaining why).
- ⏩ :  Check is not applicable to function (skip).
- 🆗 :  Check item is already implemented and does not require any edits.

#### CHECKS 📑:
1. - [ ] ✅:  Remove all lambda and direct bindings for the backend functions in:
       - [ ] 🆗: [ivy/functional/backends/jax/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/jax/general.py).
       - [ ] 🆗: [ivy/functional/backends/numpy/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/numpy/general.py).
       - [ ] 🆗: [ivy/functional/backends/tensorflow/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/tensorflow/general.py).
       - [ ] ✅: [ivy/functional/backends/torch/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/torch/general.py).
2. - [ ] 🆗: Implement the following if they don't exist:
       1. - [ ]  🆗: The `ivy.Array` instance method in [ivy/data_classes/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general.py).
       2. - [ ]  🆗: The `ivy.Array` special method in [ivy/data_classes/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).
       3. - [ ]  ⏩: The `ivy.Array` reverse special method in [ivy/data_classes/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).
       4. - [ ] 🆗: The `ivy.Container` static method in [ivy/data_classes/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).
       5. - [ ] 🆗: The `ivy.Container` instance method in [ivy/data_classes/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).
       6. - [ ]  🆗:  The `ivy.Container` special method in [ivy/data_classes/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).
       7. - [ ]  ⏩: The `ivy.Container` reverse special method in [ivy/data_classes/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).
3. - [ ] 🆗:  Make sure that the aforementioned methods are added into the correct category-specific parent class, such as  `ivy.ArrayWithElementwise`,  `ivy.ContainerWithManipulation`  etc.
4. - [ ] 🆗:  Correct all of the  [Function Arguments and the type hints](https://unify.ai/docs/ivy/overview/deep_dive/function_arguments.html#function-arguments) for every function  **and**  its  _relevant methods_, including those you did not implement yourself.
5. - [ ] 🆗: Add the correct  [Docstrings](https://unify.ai/docs/ivy/overview/deep_dive/docstrings.html#docstrings)  to every function  **and**  its  _relevant methods_, including those you did not implement yourself. The following should be added:
       1. - [ ] 🆗:   <a name="ref1"></a> The function's [Array API standard](https://data-apis.org/array-api/latest/index.html) description in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py). If the function is not part of the Array API standard then a description of similar style should be added to the same file.
	The following modifications should be made to the description:
              - [ ] 🆗:  Remove type definitions in the `Parameters` and `Returns` sections.
              - [ ] 🆗:  Add `out` to the `Parameters` section if function accepts an `out` argument.
              - [ ] 🆗:  Replace `out` with `ret` in the `Returns` section.
       2. - [ ] 🆗:  Reference to docstring for ivy.function_name ([5.a](#ref1)) for the function description **and** modified `Parameters` and `Returns` sections as described in [the docs](https://unify.ai/docs/ivy/overview/deep_dive/docstrings.html#docstrings) in:
              - [ ] 🆗:  [ivy/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general.py).
              - [ ] 🆗:  [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py) (in the static and instance method versions).
              - [ ] ⏩:   [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py) if the function has a special method  ( like `__function_name__` ).
              - [ ] ⏩:  [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py) if the function has a reverse special method  ( like `__rfunction_name__` ).
              - [ ] ⏩: [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py) if the function has a special method ( like `__function_name__` ).
              - [ ] ⏩:  [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py) if the function has a reverse special method  ( like `__rfunction_name__` ).
6. - [ ] 🆗: Add thorough  [Docstring Examples](https://unify.ai/docs/ivy/overview/deep_dive/docstring_examples.html#docstring-examples)  for every function  **and**  its  _relevant methods_  and ensure they pass the docstring tests.

		**Functional Examples** in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py).

		1. - [ ] 🆗: Cover all possible variants for each of the arguments independently (not combinatorily).
	 	2. - [ ] 🆗: Vary the values and input shapes considerably between examples.
	 	3. - [ ] 🆗: Start out simple and get more complex with each example.
	 	4. - [ ] 🆗: Show an example with:
			   - [ ] 🆗: `out` unused.
			   - [ ] 🆗: `out` used to update a new array y.
			   - [ ] 🆗: `out` used to inplace update the input array x (if x has the same dtype and shape as the return).
	 	5. - [ ] 🆗: If broadcasting is relevant for the function, then show examples which highlight this.

		**Nestable Function Examples** in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py).
		Only if the function supports nestable operations.

	 	6. - [ ] 🆗: <a name="ref2"></a> Add an example that passes in an  `ivy.Container`  instance in place of one of the arguments.
	 	7. - [ ] 🆗: <a name="ref3"></a> Add an example passes in  `ivy.Container`  instances for multiple arguments.

		**Container Static Method Examples** in [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).

	 	8. - [ ] 🆗: The example from point ([6.f](#ref2)) should be replicated, but added to the  `ivy.Container`  **static method** docstring in with  `ivy.<func_name>`  replaced with  `ivy.Container.static_<func_name>`  in the example.
	 	9. - [ ] 🆗: The example from point ([6.g](#ref3)) should be replicated, but added to the  `ivy.Container`  **static method** docstring, with  `ivy.<func_name>`  replaced with  `ivy.Container.static_<func_name>`  in the example.

		**Array Instance Method Example** in [ivy/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general).

		10. - [ ] 🆗: Call this instance method of the  `ivy.Array`  class.

		**Container Instance Method Example** in [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).

		11. - [ ] 🆗: Call this instance method of the  `ivy.Container`  class.

		**Array Operator Examples** in [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).

		12. - [ ] ⏩: Call the operator on two  `ivy.Array`  instances.
	 	13. - [ ] ⏩: Call the operator with an  `ivy.Array`  instance on the left and  `ivy.Container`  on the right.

		**Array Reverse Operator Example** in [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).

		14.  - [ ] ⏩: Call the operator with a  `Number`  on the left and an  `ivy.Array`  instance on the right.

		**Container Operator Examples** in [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).

		15. - [ ] ⏩: Call the operator on two `ivy.Container` instances containing Number instances at the leaves.
	 	16. - [ ] ⏩: Call the operator on two `ivy.Container` instances containing `ivy.Array` instances at the leaves.
	 	17. - [ ] ⏩: Call the operator with an `ivy.Container` instance on the left and `ivy.Array` on the right.

		**Container Reverse Operator Example** in [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).

		18. - [ ] ⏩: Following example in the [`ivy.Container.__radd__`](https://github.com/unifyai/ivy/blob/e28a3cfd8a4527066d0d92d48a9e849c9f367a39/ivy/container/container.py#L173) docstring, with the operator called with a `Number` on the left and an `ivy.Container` instance on the right.

		**Tests**

		19. - [ ] ✅: Docstring examples tests passing.
		20. - [ ] ✅: Lint checks passing.

**Comment by Kacper-W-Kozdon** [comment]
@AnnaTz the pull request I was asked to tag you in- for torch gather. **_Pytest might still throw some issues_** for edge cases (mostly related to dimensionality), I'm still fixing them, but it seems that I've reached the point where it's all just related to finding the right conditions for reshaping and expanding params and indices before feeding them into the native torch.gather().

**Comment by Kacper-W-Kozdon** [comment]
@AnnaTz ready to review! Probably even to merge (no errors), but due to my previous experiences with pytest for some reason failing to detect serious issues, I'd greatly appreciate it if you double-checked. _However_, the function could probably use some polishing to make it simpler, there _might_ be some redundant conditions still- and that could probably use a look as well. On the plus side- ivy.gather() on torch backend finally uses torch methods instead of borrowing from some other libraries. The results of the command `pytest -s -v --full-trace --debug=pytestdebug.log ivy_tests/test_ivy/test_functional/test_core/test_general.py::test_gather[cpu-torch-False-False]` are attached in the pytestdebug.log. 
[pytestdebug.log](https://github.com/unifyai/ivy/files/13757350/pytestdebug.log)

**Comment by vedpatwardhan** [comment]
Hey @saeedashrraf, could you please take a look at this PR, given that it's about gather? Thanks 😄

**Comment by Kacper-W-Kozdon** [comment]
> Hey @saeedashrraf, could you please take a look at this PR, given that it's about gather? Thanks 😄

I'm still removing some issues in the code but I'll appreciate any advice :) And there's still a possibility that it's not worth in this case to try to do it the way I'm trying and instead it might be better to revert gather to the version from some older commit. However, I still would like to get that piece of code to pass all the tests.

**Comment by Kacper-W-Kozdon** [comment]
@saeedashrraf the PR is ready for review.

**Comment by Kacper-W-Kozdon** [comment]
> Hey @saeedashrraf, could you please take a look at this PR, given that it's about gather? Thanks 😄

Hey @vedpatwardhan - is there a chance for a review :) ? It seems that the person assigned to review this PR has not been active throughout the last month and the PR is *probably* ready to be merged (but a check through the code and a separate pytest would definitely not hurt).

**Comment by vedpatwardhan** [comment]
> Hey @vedpatwardhan - is there a chance for a review :) ? It seems that the person assigned to review this PR has not been active throughout the last month and the PR is _probably_ ready to be merged (but a check through the code and a separate pytest would definitely not hurt).

Hey @Kacper-W-Kozdon, really sorry for the delay in reviewing. @Ishticode could you please take a quick look at the changes made if we can merge it? Thanks 😄

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hi @Kacper-W-Kozdon Hi @vedpatwardhan 
Thank you so much for all the effort. Will keep your PR on priority.
It fixes several issues but i ran `test_gather` with 200 examples on our torch backend from your branch. It seems to fail with following value mismatch.
```
E       AssertionError:  the results from backend torch and ground truth framework tensorflow do not match
E        [1.]!=[0.9995117] 
E       
E       The mismatching elements are at `False` indices:
E       
E       [False] 
E       
E       
E       Falsifying example: test_gather(
E           backend_fw='torch',
E           on_device='cpu',
E           params_indices_others=(['float16', 'int32'],
E            array([-1.], dtype=float16),
E            array([[0, 0, 0],
E                   [0, 0, 0]], dtype=int32),
E            0,
E            0),
E           fn_name='gather',
E           test_flags=FunctionTestFlags(
E               ground_truth_backend='tensorflow',
E               num_positional_args=2,
E               with_out=False,
E               with_copy=False,
E               instance_method=False,
E               test_gradients=True,
E               test_trace=False,
E               transpile=False,
E               as_variable=[False],
E               native_arrays=[False],
E               container=[False],
E               precision_mode=False,
E               test_cython_wrapper=False,
E           ),
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.78.1', b'AXicY2NkAAMoBaEZGRgZGZgYkABYGAACCAAQ') as a decorator on your test case
```
Do you have any ideas where this is emerging from and are you able to fix? Thank for the effort and let me know if something is unclear.

**Comment by Kacper-W-Kozdon** [comment]
> @reproduce_failure('6.78.1', b'AXicY2NkAAMoBaEZGRgZGZgYkABYGAACCAAQ')

Initially- I don't see anything that would cause a problem; I checked in a jupyter notebook  (https://github.com/Kacper-W-Kozdon/notebook-testing-ivy/blob/main/Testing_Python.ipynb), under the evaluated input In[3]. I'll check with the decorator and in the ivy_dev venv, but at the very least it's nothing obvious and at least the notebook doesn't throw any errors.

File: ivy/functional/backends/torch/general.py
Comment (Kacper-W-Kozdon) [reply ivy/functional/backends/torch/general.py]: Done, well spotted. 👍

**Comment by Kacper-W-Kozdon** [comment]
> Hi @Kacper-W-Kozdon Hi @vedpatwardhan Thank you so much for all the effort. Will keep your PR on priority. It fixes several issues but i ran `test_gather` with 200 examples on our torch backend from your branch. It seems to fail with following value mismatch.
> 
> ```
> E       AssertionError:  the results from backend torch and ground truth framework tensorflow do not match
> E        [1.]!=[0.9995117] 
> E       
> E       The mismatching elements are at `False` indices:
> E       
> E       [False] 
> E       
> E       
> E       Falsifying example: test_gather(
> E           backend_fw='torch',
> E           on_device='cpu',
> E           params_indices_others=(['float16', 'int32'],
> E            array([-1.], dtype=float16),
> E            array([[0, 0, 0],
> E                   [0, 0, 0]], dtype=int32),
> E            0,
> E            0),
> E           fn_name='gather',
> E           test_flags=FunctionTestFlags(
> E               ground_truth_backend='tensorflow',
> E               num_positional_args=2,
> E               with_out=False,
> E               with_copy=False,
> E               instance_method=False,
> E               test_gradients=True,
> E               test_trace=False,
> E               transpile=False,
> E               as_variable=[False],
> E               native_arrays=[False],
> E               container=[False],
> E               precision_mode=False,
> E               test_cython_wrapper=False,
> E           ),
> E       )
> E       
> E       You can reproduce this example by temporarily adding @reproduce_failure('6.78.1', b'AXicY2NkAAMoBaEZGRgZGZgYkABYGAACCAAQ') as a decorator on your test case
> ```
> 
> Do you have any ideas where this is emerging from and are you able to fix? Thank for the effort and let me know if something is unclear.

I'm guessing that the assertion error comes from the flag `test_gradients=True`? Could it be related to `rtol` or `atol` params not being set in the test? Do these two params also get passed to the gradient test? If they are, I think that this might be the case, since the raw outputs of the gather() function seem to be ok. I don't have enough experience to suggest good values for `rtol` and `atol` but in the Deep Dive they are set as 10^(-3). I will also double check how dtypes are treated in the native torch.gather(), I don't see any reason for other torch functions to affect types- though this one _shouldn't_ either. 

I ran the test with an example decorator added between @handle_test() and def test_gather()
`@example(params_indices_others=(['float16', 'int32'],
          ivy.array([-1.]),
          ivy.array([[0, 0, 0],
           [0, 0, 0]]),
          0,
          0),
          test_flags=FunctionTestFlags(
            ground_truth_backend='tensorflow',
            num_positional_args=2,
            with_out=False,
            with_copy=False,
            instance_method=False,
            test_gradients=True,
            test_trace=False,
            transpile=False,
            as_variable=[False],
            native_arrays=[False],
            container=[False],
            precision_mode=False,
            test_cython_wrapper=False,
           ),
           fn_name="gather",
         )`,
however, I got a warning: "ivy_tests/test_ivy/test_functional/test_core/test_general.py::test_gather[cpu-torch-False-False] SKIPPED (Function not implemented in backend.)", even though the test is meant to be for the functional/backend. So I'm not 100% sure this example was taken into account. The tests passed 100% (default number of examples: 25).
[pytestdebug.log](https://github.com/unifyai/ivy/files/14015591/pytestdebug.log)

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hey @Kacper-W-Kozdon 
Thank you for the suggestions. I don't think gradient tests should be disabled.
and even with tolerance added (i've set 1e-3, try to change around if you like) we get significant value mismatches. For examples one i got was
```
E       AssertionError:  the results from backend torch and ground truth framework tensorflow do not match
E        [[0.20898438 0.12158203 0.12109375 0.04736328 0.        ]
E        [0.20898438 0.12158203 0.12109375 0.04736328 0.        ]]!=[[0.0625   0.0625   0.0625   0.046875 0.      ]
E        [0.0625   0.0625   0.0625   0.046875 0.      ]]
```


Could you please try `test_gather` with more examples by adding `--num-examples 200`.

Pls let me know further if something is unclear. Thank you very much :)

**Comment by Kacper-W-Kozdon** [comment]
> Hey @Kacper-W-Kozdon 
> Thank you for the suggestions. I don't think gradient tests should be disabled.
> and even with tolerance added (i've set 1e-3, try to change around if you like) we get significant value mismatches. For examples one i got was
> ```
> E       AssertionError:  the results from backend torch and ground truth framework tensorflow do not match
> E        [[0.20898438 0.12158203 0.12109375 0.04736328 0.        ]
> E        [0.20898438 0.12158203 0.12109375 0.04736328 0.        ]]!=[[0.0625   0.0625   0.0625   0.046875 0.      ]
> E        [0.0625   0.0625   0.0625   0.046875 0.      ]]
> ```
> 
> 
> Could you please try `test_gather` with more examples by adding `--num-examples 200`.
> 
> Pls let me know further if something is unclear. Thank you very much :)

Got it, I'll update my packages (my hypothesis was behind) and rerun the tests with more examples. Once I see the details of the errors and the inputs, I'll try to isolate the part of the code which might be causing issues and let you know.

**Comment by Kacper-W-Kozdon** [comment]
> Hey @Kacper-W-Kozdon Thank you for the suggestions. I don't think gradient tests should be disabled. and even with tolerance added (i've set 1e-3, try to change around if you like) we get significant value mismatches. For examples one i got was
> 
> ```
> E       AssertionError:  the results from backend torch and ground truth framework tensorflow do not match
> E        [[0.20898438 0.12158203 0.12109375 0.04736328 0.        ]
> E        [0.20898438 0.12158203 0.12109375 0.04736328 0.        ]]!=[[0.0625   0.0625   0.0625   0.046875 0.      ]
> E        [0.0625   0.0625   0.0625   0.046875 0.      ]]
> ```
> 
> Could you please try `test_gather` with more examples by adding `--num-examples 200`.
> 
> Pls let me know further if something is unclear. Thank you very much :)

I ran the test with 200 examples- it passed. I repeated with 500- and I ran into a similar error. I can 100% confirm that the last test called before the failed assertions was gradient_test. Gradients run after asserting that the outputs of the gather function are correct (types and raw values). Right now I have no clue, why the gradients would fail with the direct outputs of gather _seemingly_ matching. Also, I used the @reproduce_failure decorator with the test that failed _and_ I received the message "E               hypothesis.errors.DidNotReproduce: Expected the test to raise an error, but it completed successfully.". Could you try to do it on your end as well with the test that failed to see if it can get reproduced and what the last function called was?

**Comment by Kacper-W-Kozdon** [comment]
Hey, @Ishticode , I'll try to either open a new issue or modify within this PR the test for gather (or rather- the helper function used in it) to limit the values in the input array to what can be correctly represented in a given dtype. Current test uses a helper function which does not have arguments to control the range of values on the input but it calls `dtypes_and_values()`, which has such arguments. It seems that this (when fed into test for gradients) might cause the errors rather than the `gather()` function itself.

**Comment by Kacper-W-Kozdon** [comment]
I'd need the #28168 PR reviewed and merged first in order to solve the issue of some tests failing due to the mismatch in values and dtype (values too big for the dtype to retain the necessary precision). Based on the feedback from Ved, it _should_ be enough to clear all the tests, given that the errors show up in the gradient test associated with test_gather through the flag in test_gather.

**Comment by Kacper-W-Kozdon** [comment]
Ready for review (and merge) together with https://github.com/unifyai/ivy/pull/28168

**Comment by Kacper-W-Kozdon** [comment]
@Ishticode would you mind taking look at https://github.com/unifyai/ivy/pull/28168 alongside this PR? The tests shouldn't throw any more errors when these two are merged.

**Review by Ishticode — COMMENTED** [review]
Hi @Kacper-W-Kozdon 
Thnx for the fixes. The tests all look fine. You really have put a lot of effort into this :)
I might ask @vedpatwardhan to have a quick look before merging. 
Sorry for the delays on my end. One question I still have is that your initial comment you said the torch backend implementation was wrong and pytest wasn't catching it. As long as i can see we have not broadened the test to include newer cases. We have only restricted it somewhat with safety and tolerances. So how do we know that those cases that were previously undetected are still not going unnoticed?

**Comment by Kacper-W-Kozdon** [comment]
> Hi @Kacper-W-Kozdon Thnx for the fixes. The tests all look fine. You really have put a lot of effort into this :) I might ask @vedpatwardhan to have a quick look before merging. Sorry for the delays on my end. One question I still have is that your initial comment you said the torch backend implementation was wrong and pytest wasn't catching it. As long as i can see we have not broadened the test to include newer cases. We have only restricted it somewhat with safety and tolerances. So how do we know that those cases that were previously undetected are still not going unnoticed?

I'm not fully sure but I think my pytest was not installed properly at that time- gather() was my first contribution, the tests were passing even though they shouldn't have, which I haven't realised at first (after I reinstalled everything, they were, indeed, throwing errors).

File: ivy/functional/backends/torch/general.py
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/general.py:190]: Just a small point here, wouldn't `axis % len(params.shape)` always be positive? I suppose we wouldn't need the `if` check in that case?

Comment (Kacper-W-Kozdon) [reply ivy/functional/backends/torch/general.py:190]: I'll check that but I think there was some issue with 'torch.expand()' in the batching part or in slicing- or some other element, I don't remember right now. I only remember that I added that because of the batching in general.

File: ivy/functional/backends/torch/general.py
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/general.py:196]: not sure why we're detaching the indices here, probably it's an integer tensor anyway right? Sorry to be a pain, but given that this is a very important function, we'd need to be quite careful before merging any changes to it 😅

Comment (Kacper-W-Kozdon) [reply ivy/functional/backends/torch/general.py:196]: Not a problem at all- in this case, it's probably unnecessary, but as quite often with tensors/arrays, it's easy to accidentally alter the input which is meant to be unchanged, so when in doubt, I write the code with cloning first. In this case, the cloning can probably be dropped whatsoever and `indices_ex` can probably get replaced with just `indices` within that function. I'll do it and check if nothing breaks. Thanks for pointing that out :)

Comment (Kacper-W-Kozdon) [reply ivy/functional/backends/torch/general.py]: @vedpatwardhan Changes applied, tests passed for `num_examples=200` but I am not discouraging you from running pytest separately :)

**Review by vedpatwardhan — APPROVED** [review]
lgtm! Could you please revert the changes to the `demos` submodule? The PR should be good to merge once that's done @Ishticode. Thanks @Kacper-W-Kozdon 😄

**Comment by Kacper-W-Kozdon** [comment]
> lgtm! Could you please revert the changes to the `demos` submodule? The PR should be good to merge once that's done @Ishticode. Thanks @Kacper-W-Kozdon 😄

Done, docs/demos changes reverted 👍

**Comment by Kacper-W-Kozdon** [comment]
@vedpatwardhan @Ishticode I added explicitly the tolerance for `bfloat16`- hopefully, this will prevent the gradient test tripping the tolerances assertion on this type. Other than that, it seems to be ready.

**Review by vedpatwardhan — APPROVED** [review]
Would be worth confirming if the failures in [`test_paddle_gather`](https://github.com/unifyai/ivy/actions/runs/8263856194/job/22606560036?pr=27757#step:6:26) and [`test_paddle_gather_nd`](https://github.com/unifyai/ivy/actions/runs/8263856194/job/22606560036?pr=27757#step:6:37) are related to the changes made. Otherwise we're good to merge, thanks @Kacper-W-Kozdon @Ishticode 😄

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Review by Ishticode — APPROVED** [review]
the two tests referenced above were failing previously already. I think its good to merge. Thank you very much @Kacper-W-Kozdon for the great effort into this contribution. Thank you very much

**Pull request merged**
